### PR TITLE
add: test instructions icon animation

### LINKED ISSUE #9399: [a11y] hover/active state on "test mode" instructions
### Describe the bug

"Test mode" instructions are missing hover/active states when hovering/clicking (respectively) on the button to copy the test number.
"Test mode" instructions are also missing dark theme support.

### To Reproduce

1. Enable "test mode"
2. Go to checkout
3. Click on the card payment method

### Actual behavior

Hover/click states on the clickable test number are not present.

### Screenshots

359816988-2d3e32eb-f476-44ef-98d4-16190ac00e3f (1)

### Expected behavior

359859743-0f4d45c4-2ac4-45b6-826f-313a54caa554

### PR DESCRIPTION
**Pull request by frosso**
Fixes https://github.com/Automattic/woocommerce-payments/issues/9399

#### Changes proposed in this Pull Request

Adding a little animation on hover, click, success of the clickable payment method icon - as discussed here: pfHfG4-gT-p2#comment-238

I needed to change the markup being generated:
- Before: button and test number were two separate elements.
- Now: button wraps the test number, to make the click target bigger.

Demo on blocks and cart checkout:

Also tested on a bunch of other theme variations (TT4, Storefront, TT3, TT2 - blocks & shortcode checkout):
Screenshot 2024-08-21 at 12 02 17 PM
Screenshot 2024-08-21 at 12 02 13 PM
Screenshot 2024-08-21 at 12 01 54 PM
Screenshot 2024-08-21 at 12 01 45 PM
Screenshot 2024-08-21 at 11 48 38 AM
Screenshot 2024-08-21 at 11 46 18 AM

#### Testing instructions

- Ensure WooPayments is in test mode
- Go to the checkout (you can test both block and shortcode)
- Clicking on the test number should copy the test number
- Tested for card, Becs, and SEPA

-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `9290` or branch name `update/test-instructions-icon-animation` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/test-instructions-icon-animation) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 559fd3585e2f913e3f4425f5813dee64e453473d
- Build time: 2024-09-06 12:21:35 UTC

_Note: the build is updated when a new commit is pushed to this PR._

File: includes/payment-methods/class-becs-payment-method.php [hunk 1]
```diff
@@ -39,6 +39,6 @@ public function __construct( $token_service ) {
 	 * @return string
 	 */
 	public function get_testing_instructions() {
-		return __( '<strong>Test mode:</strong> use the test account number <number>000123456</number><button></button>. Other payment methods may redirect to a Stripe test page to authorize payment. More test card numbers are listed <a>here</a>.', 'woocommerce-payments' );
```
Comment (frosso) [inline includes/payment-methods/class-becs-payment-method.php:42]: I needed to change the markup being generated.
Before: button and test number were two separate elements.
Now: button wraps the test number, to make the click target bigger.

File: client/checkout/style.scss [hunk 2]
```diff
@@ -3,27 +3,46 @@
 }
 
 .js-woopayments-copy-test-number {
-	line-height: 1.2em;
```
Comment (frosso) [inline client/checkout/style.scss]: Since the markup changed, the styles needed to change as well.
Before, the `button` element didn't have any text, so we didn't really care what it looked like.
Now, we need to ensure the text within the `button` element plays nice across different themes and browsers.

So I reset the `padding`, changed the `font-size` and `font-weight`, ensured the `cursor` was set to `pointer` (I dunno why, but it wasn't before by default), and text color and background color matched the surrounding text.

I tested this across different themes, but I'm not overly concerned about compatibility.
This is only applied when the plugin is in test mode, so in production the whole instructions set is not going to be visible.

File: client/checkout/style.scss [hunk 3]
```diff
@@ -9,25 +9,46 @@
 	border: none !important;
 	// some themes might override the color on all `button`s - whose selector has higher specificity and our class selector.
 	background-color: transparent !important;
-	padding: 3px;
+	font-weight: normal;
```
Comment (frosso) [inline client/checkout/style.scss:12]: Since the markup changed, the styles needed to change as well.
Before, the `button` element didn't have any text, so we didn't really care what it looked like.
Now, we need to ensure the text within the `button` element plays nice across different themes and browsers.

So I reset the `padding`, changed the `font-size` and `font-weight`, ensured the `cursor` was set to `pointer` (I dunno why, but it wasn't before by default), and text color and background color matched the surrounding text.

I tested this across different themes, but I'm only somewhat concerned about compatibility.
This is only applied when the plugin is in test mode, so in production the whole instructions set is not going to be visible.

File: client/checkout/style.scss [hunk 4]
```diff
@@ -9,25 +9,46 @@
...
+	padding: 2px 1px;
+	align-items: center;
+
+	span {
+		margin-right: 4px;
+	}
 
 	i {
-		display: inline-block;
+		order: 1;
```
Comment (mdmoore) [inline client/checkout/style.scss:25]: I'm sure there's a good reason for this, but I'm curious why the `<i>` tag couldn't be placed after the span in the HTML rather than ordering it here.

File: client/checkout/utils/copy-test-number.js [hunk 5]
```diff
@@ -41,6 +42,12 @@ document.addEventListener(
 					context: 'wc/checkout/payments',
 				}
 			);
+		copyNumberButton.classList.remove( 'state--success' );
```
Comment (mdmoore) [inline client/checkout/utils/copy-test-number.js]: Similar to the previous comment, I'm sure there's a reason, but curious why we're removing the check before it's added? I don't see the class actually being removed.

**Review by mdmoore — APPROVED** [review]
I left a couple questions, but this works well as is! Looks good on the various themes mentioned and works well with keyboard navigation. Thanks @frosso!

File: client/checkout/style.scss (hunk 4 above)
Comment (frosso) [reply client/checkout/style.scss:25]: I tried to do that in the `get_testing_instructions()` pattern, but the replacement done by `WC_Payments_Utils::esc_interpolated_html()` never really worked for me :( 
it looks like `WC_Payments_Utils::esc_interpolated_html` attempts to close any opened HTML tag. I couldn't figure out how to open an `<i>` after the `<button>`'s open tag, but before the `<button>`'s tag got closed :suboptimal:

Let me know if you have a better solution! And sorry I missed clarifying this in a comment, it was quite tricky to get the markup right :D

File: client/checkout/utils/copy-test-number.js (hunk 5 above)
Comment (frosso) [reply client/checkout/utils/copy-test-number.js]: I'm doing it just in case the button is clicked multiple times in quick sequence, to avoid adding the `state--success` class name multiple times

Comment (mdmoore) [reply client/checkout/utils/copy-test-number.js]: I thought that might be the case. I skipped this line and tried clicking multiple times on the button but don't see the class added multiple times. I think `.add` doesn't add the class if it's already there which I didn't realize. 

>https://developer.mozilla.org/en-US/docs/Web/API/DOMTokenList/add
>adds the given tokens to the list, omitting any that are already present.

I actually don't see it being removed by that line at all which is strange. I know it happens quickly, but there would normally be an indication/highlight/flash on the element in the inspector that class was removed.

Anyways, fine to leave it if you are seeing something different!

File: client/checkout/style.scss (hunk 4 above)
Comment (mdmoore) [reply client/checkout/style.scss:25]: Ah, ok. Yeah I see how it's a pain. Would this work? 

```
// class-cc-payment-method.php - line 72
public function get_testing_instructions() {
	return __( '<strong>Test mode:</strong> use test card <button><span>4242 4242 4242 4242</span><i></i></button> or refer to our <a>testing guide</a>.', 'woocommerce-payments' );
}
```

```
// class-wc-payments-checkout.php - line 326
$settings[ $payment_method_id ]['testingInstructions'] = WC_Payments_Utils::esc_interpolated_html(
	/* translators: link to Stripe testing page */
	$payment_method->get_testing_instructions(),
	[
		'a'      => '<a href="https://woocommerce.com/document/woopayments/testing-and-troubleshooting/testing/#test-cards" target="_blank">',
		'strong' => '<strong>',
		'button' => '<button type="button" class="js-woopayments-copy-test-number" aria-label="' . esc_attr( __( 'Click to copy the test number to clipboard', 'woocommerce-payments' ) ) . '" title="' . esc_attr( __( 'Copy to clipboard', 'woocommerce-payments' ) ) . '">',
		'i'      => '<i>',
		'span'   => '<span>',
	]
);
```

Comment (frosso) [reply client/checkout/style.scss:25]: Yes, that works - however, I wanted to ensure the markup in the translatable string was kept to a minimum to simplify the translators' job. It would be easy to use the wrong markup and not have the correct result in the interpolation. By just having the `<button>` tag, we're limiting the risk of using the wrong markup in case the translators don't understand how the HTML tags don't fit into the string.

In short, I was trying to simplify the HTML in the string - but I can be persuaded :D what do you think?

File: client/checkout/utils/copy-test-number.js (hunk 5 above)
Comment (frosso) [reply client/checkout/utils/copy-test-number.js]:
```suggestion
```

Yeah, I think you're right - I tested it in multiple browsers myself, and it doesn't seem to be added multiple times.
Removing the `.remove()` :)

File: client/checkout/style.scss (hunk 4 above)
Comment (mdmoore) [reply client/checkout/style.scss:25]: That's fair and I think more important than the order of the elements. Thanks for explaining!
